src/coprocessor/codec/mysql/duration.rs

```diff
@@ -253,73 +239,75 @@ mod parser {
 
 } /* parser */
 
-/// `Duration` is the type for MySQL `time` type.
-///
-/// It only occupies 24 bytes in memory.
-/// TODO: reduce size of `Duration` to 8 bytes
-#[derive(Debug, Clone, Copy)]
-pub struct Duration {
-    dur: StdDuration,
-    /// Fractional Seconds Precision
-    /// See http://dev.mysql.com/doc/refman/5.7/en/fractional-seconds.html
-    fsp: u8,
-    neg: bool,
+bitfield! {
+    #[derive(Clone, Copy)]
+    pub struct Duration(u64);
+    impl Debug;
+    #[inline]
+    pub bool, neg, set_neg: 55;
+    #[inline]
+    bool, unused, _: 54;
+    #[inline]
+    pub u32, hours, set_hours: 53, 44;
+    #[inline]
+    pub u32, minutes, set_minutes: 43, 38;
+    #[inline]
+    pub u32, secs, set_secs: 37, 32;
+    #[inline]
+    u32, micros, set_micros: 31, 8;
+    #[inline]
+    pub u8, fsp, set_fsp: 7, 0;
 }
 
 impl Duration {
+    /// return the identity element of `Duration`
     pub fn zero() -> Duration {
-        Duration {
-            dur: StdDuration::from_secs(0),
-            neg: false,
-            fsp: 0,
-        }
-    }
-
-    pub fn fsp(&self) -> u8 {
-        self.fsp
+        Duration(0)
     }
 
-    /// Changes the FSP part of the duration.
-    pub fn set_fsp(&mut self, fsp: u8) {
-        self.fsp = fsp;
+    pub fn is_zero(mut self) -> bool {
+        self.set_neg(false);
+        self.set_fsp(0);
+        self.0 == 0
     }
 
-    pub fn hours(&self) -> u64 {
-        self.dur.as_secs() / SECS_PER_HOUR
+    /// Return the absolute value of `Duration`
+    pub fn abs(mut self) -> Self {
+        self.set_neg(false);
+        self
     }
 
-    pub fn minutes(&self) -> u64 {
-        self.dur.as_secs() % SECS_PER_HOUR / SECS_PER_MINUTE
+    /// return the fractional part of `Duration`, in whole nanoseconds.
+    pub fn subsec_nanos(self) -> u32 {
```

Comment:
Since we only provide microsecond precision, I prefer to just remove this function.